src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -454,11 +462,18 @@ void ClusterTaskManager::TasksUnblocked(const std::vector<TaskID> &ready_ids) {
 void ClusterTaskManager::TaskFinished(std::shared_ptr<WorkerInterface> worker,
                                       RayTask *task) {
   RAY_CHECK(worker != nullptr && task != nullptr);
+  if (leased_workers_.count(worker->WorkerId()) == 0) {
```

Comment:
why is count a leaky abstraction here? i'm mostly worried that we don't consistently unset these things (we don't event have that op for task id and task), we rely on removing them from the map so it would be easy to miss edge cases if we tried that.

Reply:
What I meant is that this is checking whether the worker is in the leased workers map, but this class doesn't actually erase from that map. So the logic here is not self-contained and e2e correctness depends on an implicit assumption about the caller (brittle).

Put another way, we want to write this in a way that is unit-testable. Currently, it's not because there's no way to test that the NodeManager is updating leased_workers_ and calling this method in the correct order.

Reply:
I mean I generally agree with those statements, but I don't understand why that's in scope of this PR. 

If that's a refactor that's worth doing, then I'm all for it, but I don't see why _this_ PR is the time to make that change since we've been sharing the state of `leased_workers_` since the new scheduler was written to begin with.

Reply:
The problem is that this code was previously idempotent (I believe releasing task resources is idempotent), but now it's not because of the new task count. So we need to make sure it's idempotent in a way that is still unit-testable.

I don't think this is a terribly big change, as it just requires unsetting the task spec assigned to the worker. Is there a problem with that solution?